### DISCUSSION TITLE
Help with a docker compose set up

### DISCUSSION THREAD

**Discussion by pedz** [emacs-lsp/lsp-docker#59] [Q&A] [answered]
I have a docker compose with two containers: db and web.  I am developing a Ruby on Rails app.  It appears as if there are currently two choices to connect to a language server.  Either I can have my language server in another image and use `docker run` or I can have it in a stopped container and use `docker start`.

I think what I’d like in my case instead is to simply do:
```sh
   docker exec -it app-web-1 /root/bin/solargraph stdio
```
because the `docker compose up` has already created the two containers needed.  I have solargraph in my Gemfile so the only question is how to do that.  I say “I think” because I’m still groping around trying to figure out how this stuff is intended to work.

I see there is a `lsp-docker-exec-in-container` function but it isn’t used anywhere.

I’m tempted to hack on `lsp-docker-register` and create a third option that uses `lsp-docker-exec-in-container` rather than `lsp-docker-launch-existing-container` or `lsp-docker-launch-new-container`.  I’m likely to try that out and see if that works but I thought I would also ask for help.

**Comment by yyoncho** [comment]
If I am reading this correctly, you want to run the server in existing container, right? Then lsp-docker-launch-existing-container seems to be doing what you want? (I haven't touched this code base for a lot of time)

**Reply by pedz** [reply]
lsp-docker-launch-existing-container does `docker start ...`  I may not be using `docker compose` correctly.  I'm breaking the "one service per container" rule I bet.

The web container exists but is already running the rails app in server mode.  I think I want the solargraph server to be running in the same container because that container has all of the gem files but I suppose I could just use the same image since everything is either in the image or on the local file system.

Let me explore that avenue and see if I can get things to work.

**Comment by pedz** [comment]
So far, I can not get lsp-mode to call anything within lsp-docker.el.  I asked a question on the [Emacs Stack Exchange](https://emacs.stackexchange.com/q/73872/14815).  I see such things as `lsp-docker-start` and `lsp-docker-register` (one calls the other) but the first test is for `(lsp-workspace-root)` which is nil so nothing happens.

I'm trying to figure out how to call `lsp-docker-init-clients` but so far it isn't obvious.

There are a lot of mysterious "id"s like server-id, client-id, etc, which I'm trying to puzzle out what they mean and where they come from.

**Comment by pedz** [comment, marked as answer]
I wrote some lisp code for my particular set up.
```lisp
(require 'lsp-docker)

;; In the lsp-mode directory (where lsp-mode.el lives) is a
;; subdirectory called "clients".  In there is a list of "clients".  I
;; can't figure these terms out BUT the list below is a list of these
;; "clients" that you want to be enabled inside the client / server /
;; THING that you are defining / creating.  In my case, as far as I
;; know, I want just the Solargraph "client".
;;
;; Actually this list is just a list of packages to `require' at the
;; time the THING is registered -- so... it is much ado about nothing?
(setq lsp-docker-client-packages
    '(lsp-solargraph))

;; From reading, I guess they call the external server the "server"
;; and the piece either inside or outside of Emacs talking to the
;; server, the "client".  But it gets very confusing (keep reading).
;;
;; Emacs has structures which are defined via `cl-defstruct'.
;; `lsp-clients' is a hash of `lsp--client' structures which has a
;; "slot" `server-id' whose comment is: "Unique identifier for
;; representing the client object.".
;;
;; As mentioned `lsp-clients' is a hash (key value pairs) with the
;; keys being the `server-id' (which is also within the structure).
;;
;; With all that in mind, then:
;;
;; :server-id -- needs to be an existing entry in lsp-clients.  You
;;     can see the list and their priorities by executing:
;;
;;         (maphash
;;          (lambda (key value)
;;            (insert (format "%S  %d\n"
;;                            key
;;                            (lsp--client-priority value))))
;;          lsp-clients)
;;
;;     in the *scratch*  buffer.
;;
;; :docker-server-id -- The name or id of the new "client object"
;;     a.k.a. the `server-id'.  The existing entry is found via the
;;     `server-id', updated, and entered as a new client using the
;;     docker-server-id.
;;
;; :server-command -- This string is munged via
;;     `lsp-docker-launch-new-container' into a list of strings which
;;     is used to start the language server inside the Docker
;;     image/container.  Basically it will turn into:
;;
;;         docker run --name TBD1 --rm -i TBD2 TBD3 `server-command'
;;
;;     TBD1 is `docker-container-name'; TBD2 are `path-mappings' that
;;     have been transformed into a list of "-v" options.  TBD3 is
;;     `docker-image-id'
;;
;; `docker-container-name' and `docker-image-id' will be specified below.
;;
(setq lsp-docker-client-configs
      '((:server-id ruby-ls :docker-server-id rubyls-docker :server-command "/root/bin/solargraph stdio")))

(lsp-docker-init-clients
 :path-mappings '(("/Users/pedz/Source/hatred" . "/hatred"))
 :docker-image-id "pedzsan/hatred:1.0"
 :docker-container-name "docker-ruby-lsp"
 :priority 10
 :client-packages lsp-docker-client-packages
 :client-configs lsp-docker-client-configs)

(provide 'lsp-docker-start)
```

**Comment by pedz** [comment]
I was documenting, for myself, what I had found and started wondering.  I'll cut to the chase.  If you want to read more, you can read about it on this reddit [post](https://www.reddit.com/r/emacs/comments/ydyd95/what_dont_dired_buffers_have_a_bufferfilename/?utm_source=share&utm_medium=web2x&context=3)

I was trying to do `lsp-docker-start` from a dired buffer.  dired buffers return nil for `buffer-file-name` so `lsp-workspace-root` returns nil and the LSP server isn't started.

I've not tried it but had I tried `lsp-docker-start` from a file, I would have gotten much different results and, probably, the YAML configuration file would have worked.

I'm posting this for others to find and perhaps learn from my mistakes.

**Reply by factyy** [reply]
Thank you for such a detailed post! While I was working on persistent configurations (including using containers instead of images for language servers) I focused only on the case of registering language servers while you are in a file.

The other case you mentioned (relying on `docker-compose` to create and start containers while using `docker exec` to start the language server in a running container): I was planning to add it, but still haven't got the time to do so. If you are interested, your help is very welcome :)

P.S: keep in mind that there is `podman` out there and I don't know whether its' `podman-compose` is reliable enough
